mono/profiler/mono-profiler-log.c

```diff
@@ -652,29 +729,31 @@ buffer_unlock (void)
 {
 	mono_memory_barrier ();
 
+	gint32 state = InterlockedRead (&buffer_lock_state);
+
 	// See the comment in buffer_lock ().
-	if (InterlockedReadPointer (&buffer_rwlock_exclusive) == (gpointer) thread_id ())
+	if (state == PROF_TLS_GET ()->small_id << 16)
 		return;
 
-	g_assert (InterlockedRead (&buffer_rwlock_count) && "Why are we trying to decrement a zero reader count?");
+	g_assert (state && "Why are we decrementing a zero reader count?");
+	g_assert (!(state >> 16) && "Why is the exclusive lock held?");
 
-	InterlockedDecrement (&buffer_rwlock_count);
+	InterlockedDecrement (&buffer_lock_state);
 }
 
 // Cannot be used recursively.
 static void
 buffer_lock_excl (void)
 {
-	gpointer tid = (gpointer) thread_id ();
+	gint32 new_ = PROF_TLS_GET ()->small_id << 16;
 
-	g_assert (InterlockedReadPointer (&buffer_rwlock_exclusive) != tid && "Why are we taking the exclusive lock twice?");
+	g_assert (InterlockedRead (&buffer_lock_state) != new_ && "Why are we taking the exclusive lock twice?");
 
-	MONO_ENTER_GC_SAFE;
+	InterlockedWrite (&buffer_lock_exclusive_intent, 1);
```

Comment:
Yeah, it isn't a perfect solution. It's meant as nothing more than a suggestion to readers to back off the next time they try to increment the reader count.

That said, I believe this is not an issue in practice. We only take the writer lock when a thread is GCing and from the helper thread when doing periodic sync points. If those two things happen at the same time, it doesn't actually matter that a few readers manage to get in between them. The important thing is that is that a sync point gets emitted; whether it's by the GCing thread or the helper thread doesn't really make a big difference.